Error while checking build/O2/fullCI for e02cd4e1014ed09e1af90bcefdb5558d6fe7befc at 2024-04-02 19:22:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/SOURCES/O2/12971-slc8_x86-64/0/Detectors/EMCAL/calibration/src/EMCALPedestalHelper.cxx:104:18: error: statement should be inside braces [readability-braces-around-statements]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/12971/e02cd4e1014ed09e1af90bcefdb5558d6fe7befc/build_O2_fullCI/pretty.html).

--- a/Detectors/EMCAL/calib/include/EMCALCalib/CalibDB.h
+++ b/Detectors/EMCAL/calib/include/EMCALCalib/CalibDB.h
@@ -8,6 +8,10 @@
 // In applying this license CERN does not waive the privileges and immunities
 // granted to it by virtue of its status as an Intergovernmental Organization
 // or submit itself to any jurisdiction.
+
+#ifndef ALICEO2_EMCAL_CALIBDB
+#define ALICEO2_EMCAL_CALIBDB
+
 #include <exception>
 #include <map>
 #include <string>
@@ -379,3 +383,5 @@ class CalibDB
 } // namespace emcal
 
 } // namespace o2
+
+#endif

--- a/Detectors/EMCAL/calibration/CMakeLists.txt
+++ b/Detectors/EMCAL/calibration/CMakeLists.txt
@@ -16,6 +16,7 @@ o2_add_library(EMCALCalibration
                         src/EMCALCalibExtractor.cxx
                         src/EMCALCalibParams.cxx
                         src/EMCDCSProcessor.cxx
+                        src/EMCALPedestalHelper.cxx
                         src/PedestalCalibDevice.cxx
                         src/PedestalProcessorDevice.cxx
                         src/PedestalProcessorData.cxx
@@ -43,6 +44,7 @@ o2_target_root_dictionary(EMCALCalibration
                                   include/EMCALCalibration/EMCALTimeCalibData.h
                                   include/EMCALCalibration/EMCALCalibParams.h
                                   include/EMCALCalibration/EMCDCSProcessor.h
+                                  include/EMCALCalibration/EMCALPedestalHelper.h
                                   include/EMCALCalibration/PedestalProcessorData.h
                           LINKDEF src/EMCALCalibrationLinkDef.h)
 

--- a/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALPedestalHelper.h
+++ b/Detectors/EMCAL/calibration/include/EMCALCalibration/EMCALPedestalHelper.h
@@ -1,0 +1,65 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef EMCAL_PEDESTAL_HELPER_H_
+#define EMCAL_PEDESTAL_HELPER_H_
+
+#include "CCDB/CcdbApi.h"
+#include "CCDB/BasicCCDBManager.h"
+#include "CCDB/CCDBTimeStampUtils.h"
+#include "EMCALReconstruction/Channel.h"
+#include "EMCALBase/Geometry.h"
+#include "EMCALBase/Mapper.h"
+#include "EMCALCalib/CalibDB.h"
+#include "EMCALCalib/Pedestal.h"
+
+#include <algorithm>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+
+namespace o2::emcal
+{
+
+class EMCALPedestalHelper
+{
+
+ public:
+  EMCALPedestalHelper() = default;
+  ~EMCALPedestalHelper() = default;
+
+  /// \brief Encodes the pedestal object into a string. This function fills fMeanPed which is then converted to a string in createInstructionString
+  /// \param obj pedestal object as stored in production ccdb
+  std::vector<char> createPedestalInstruction(const Pedestal& obj);
+
+  /// \brief print the vector produced by createInstructionString in a textfile
+  void dumpInstructions(const std::string_view filename, const std::vector<char> data, int mRun);
+
+ private:
+  /// \brief initialize fMeanPed with zeros
+  void setZero();
+
+  /// \brief converts fMeanPed to a vector of char
+  std::vector<char> createInstructionString();
+
+  static constexpr short kNSM = 20;    ///< number of SuperModules
+  static constexpr short kNRCU = 2;    ///< number of readout crates (and DDLs) per SM
+  static constexpr short kNDTC = 40;   ///< links for full SRU
+  static constexpr short kNBranch = 2; ///< low gain/high gain
+  static constexpr short kNFEC = 10;   ///< 0..9, when including LED Ref
+  static constexpr short kNChip = 5;   ///< really 0,2..4, i.e. skip #1
+  static constexpr short kNChan = 16;
+  short fMeanPed[kNSM][kNRCU][kNBranch][kNFEC][kNChip][kNChan];
+};
+
+} // namespace o2::emcal
+
+#endif

--- a/Detectors/EMCAL/calibration/include/EMCALCalibration/PedestalCalibDevice.h
+++ b/Detectors/EMCAL/calibration/include/EMCALCalibration/PedestalCalibDevice.h
@@ -21,6 +21,7 @@
 #include "EMCALBase/Geometry.h"
 #include "EMCALCalibration/PedestalProcessorData.h"
 #include "EMCALCalibration/EMCALCalibExtractor.h"
+#include "EMCALCalibration/EMCALPedestalHelper.h"
 
 namespace o2::emcal
 {
@@ -28,7 +29,7 @@ namespace o2::emcal
 class PedestalCalibDevice : o2::framework::Task
 {
  public:
-  PedestalCalibDevice() = default;
+  PedestalCalibDevice(bool dumpToFile) : mDumpToFile(dumpToFile){};
   ~PedestalCalibDevice() final = default;
 
   void init(framework::InitContext& ctx) final;
@@ -45,12 +46,14 @@ class PedestalCalibDevice : o2::framework::Task
 
  private:
   Geometry* mGeometry = nullptr;                  ///< pointer to the emcal geometry class
-  o2::emcal::EMCALCalibExtractor mCalibExtractor; // instance of the calibration extraction class                                                                ///< Calibration postprocessing
-  PedestalProcessorData mPedestalData;            /// pedestal data to accumulate data
-  long int mStartTS = 0;
+  o2::emcal::EMCALCalibExtractor mCalibExtractor; ///< instance of the calibration extraction class                                                                ///< Calibration postprocessing
+  PedestalProcessorData mPedestalData;            ///< pedestal data to accumulate data
+  long int mStartTS = 0;                          ///< timestamp at the start of run used for the object in the ccdb
+  bool mDumpToFile;                               ///< if output of pedestal calib (DCS ccdb) should be written to text file
+  int mRun = 0;                                   ///< current run number
 };
 
-o2::framework::DataProcessorSpec getPedestalCalibDevice();
+o2::framework::DataProcessorSpec getPedestalCalibDevice(bool dumpToFile);
 
 } // end namespace o2::emcal
 

--- a/Detectors/EMCAL/calibration/src/EMCALPedestalHelper.cxx
+++ b/Detectors/EMCAL/calibration/src/EMCALPedestalHelper.cxx
@@ -1,0 +1,192 @@
+// Copyright 2019-2020 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#include "EMCALCalibration/EMCALPedestalHelper.h"
+using namespace o2::emcal;
+
+std::vector<char> EMCALPedestalHelper::createPedestalInstruction(const Pedestal& obj)
+{
+
+  setZero();
+
+  o2::emcal::Geometry* geo = o2::emcal::Geometry::GetInstanceFromRunNumber(300000);
+  o2::emcal::MappingHandler mapper;
+
+  for (int itower = 0; itower < 17664; itower++) {
+    auto [ddl, row, col] = geo->getOnlineID(itower);
+    auto [sm, mod, iphi, ieta] = geo->GetCellIndex(itower);
+    const auto& mapping = mapper.getMappingForDDL(ddl);
+    int ircu = ddl % 2;
+    auto addressLG = mapping.getHardwareAddress(row, col, o2::emcal::ChannelType_t::LOW_GAIN),
+         addressHG = mapping.getHardwareAddress(row, col, o2::emcal::ChannelType_t::HIGH_GAIN);
+    auto fecLG = o2::emcal::Channel::getFecIndexFromHwAddress(addressLG),
+         fecHG = o2::emcal::Channel::getFecIndexFromHwAddress(addressHG),
+         branchLG = o2::emcal::Channel::getBranchIndexFromHwAddress(addressLG),
+         branchHG = o2::emcal::Channel::getBranchIndexFromHwAddress(addressHG),
+         chipLG = o2::emcal::Channel::getAltroIndexFromHwAddress(addressLG),
+         chipHG = o2::emcal::Channel::getAltroIndexFromHwAddress(addressHG),
+         channelLG = o2::emcal::Channel::getChannelIndexFromHwAddress(addressLG),
+         channelHG = o2::emcal::Channel::getChannelIndexFromHwAddress(addressHG);
+    fMeanPed[sm][ircu][branchHG][fecHG][chipHG][channelHG] = obj.getPedestalValue(itower, false, false);
+    fMeanPed[sm][ircu][branchLG][fecLG][chipLG][channelLG] = obj.getPedestalValue(itower, true, false);
+  }
+
+  for (int iledmon = 0; iledmon < 480; iledmon++) {
+    int sm = iledmon / 24,
+        col = sm % 24,
+        ircu = 0, // LEDMONS always on RCU 0
+      iddl = 2 * sm + ircu;
+    const auto& mapping = mapper.getMappingForDDL(iddl);
+    auto addressLG = mapping.getHardwareAddress(0, col, o2::emcal::ChannelType_t::LEDMON),
+         addressHG = mapping.getHardwareAddress(1, col, o2::emcal::ChannelType_t::LEDMON);
+    auto fecLG = o2::emcal::Channel::getFecIndexFromHwAddress(addressLG),
+         fecHG = o2::emcal::Channel::getFecIndexFromHwAddress(addressHG),
+         branchLG = o2::emcal::Channel::getBranchIndexFromHwAddress(addressLG),
+         branchHG = o2::emcal::Channel::getBranchIndexFromHwAddress(addressHG),
+         chipLG = o2::emcal::Channel::getAltroIndexFromHwAddress(addressLG),
+         chipHG = o2::emcal::Channel::getAltroIndexFromHwAddress(addressHG),
+         channelLG = o2::emcal::Channel::getChannelIndexFromHwAddress(addressLG),
+         channelHG = o2::emcal::Channel::getChannelIndexFromHwAddress(addressHG);
+    fMeanPed[sm][ircu][branchHG][fecHG][chipHG][channelHG] = obj.getPedestalValue(iledmon, false, true);
+    fMeanPed[sm][ircu][branchLG][fecLG][chipLG][channelLG] = obj.getPedestalValue(iledmon, true, true);
+  }
+
+  return createInstructionString();
+}
+
+void EMCALPedestalHelper::setZero()
+{
+  for (int ism = 0; ism < kNSM; ism++) {
+    for (int ircu = 0; ircu < kNRCU; ircu++) {
+      for (int ibranch = 0; ibranch < kNBranch; ibranch++) {
+        for (int ifec = 0; ifec < kNFEC; ifec++) {
+          for (int ichip = 0; ichip < kNChip; ichip++) {
+            for (int ichan = 0; ichan < kNChan; ichan++) {
+              fMeanPed[ism][ircu][ibranch][ifec][ichip][ichan] = 0;
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+std::vector<char> EMCALPedestalHelper::createInstructionString()
+{
+  std::stringstream fout;
+
+  unsigned int lineValue = 0;
+
+  const unsigned int FECheaderCode = 0xC0000000;
+  //  const unsigned int FECwordCode   = 0x80000000;
+  const unsigned int FEClineCode = 0x40000000;
+
+  short iSM = 0;
+  short iRCU = 0;
+  short ibranch = 0;
+  short iFEC = 0;
+  short ichip = 0;
+  short ichan = 0;
+  short Ped = 0;
+  short iDTC = 0;
+
+  for (iSM = 0; iSM < kNSM; iSM++) {
+    int iside = iSM % 2;
+    int isect = iSM / 2;
+    if (iSM > 11) {
+      isect += 3; // skip non-installed sectors
+    }
+
+    int activeDTC[kNDTC] = {0};
+    for (iDTC = 0; iDTC < kNDTC; iDTC++) {
+      if (iDTC == 10 || iDTC == 20 || iDTC == 30) { // skip TRU
+        activeDTC[iDTC] = 0;
+      } else {
+        if (iSM < 10) { // not special third SMs or DCal SMs
+          activeDTC[iDTC] = 1;
+        } else {
+          if (iSM == 10 || iSM == 19) { // SMA5 or SMC12
+            if (iDTC < 14) {
+              activeDTC[iDTC] = 1;
+            } else {
+              activeDTC[iDTC] = 0;
+            }
+          } else if (iSM == 11 || iSM == 18) { // SMC5 or SMA12
+            if (iDTC == 0 || iDTC >= 27) {
+              activeDTC[iDTC] = 1;
+            } else {
+              activeDTC[iDTC] = 0;
+            }
+          } else {
+            // DCal... no FECs in  9,11-13, 23-26, 36-39
+            if ((iDTC >= 9 && iDTC <= 13) || (iDTC >= 23 && iDTC <= 26) ||
+                (iDTC >= 36 && iDTC <= 39)) {
+              activeDTC[iDTC] = 0;
+            } else {
+              activeDTC[iDTC] = 1;
+            }
+          } // DCal
+        }   // non-EMCal
+      }     // non-TRU
+    }
+
+    // OK, let's generate the files for all active FECs/DTCs
+    for (iDTC = 0; iDTC < kNDTC; iDTC++) {
+      if (activeDTC[iDTC] == 0) {
+        continue;
+      }
+
+      lineValue = FECheaderCode | isect << 9 | iside << 8 | iDTC;
+      fout << lineValue << std::endl;
+
+      iRCU = iDTC / 20;
+      ibranch = (iDTC % 20) / 10;
+      iFEC = iDTC % 10;
+      int ipos = iFEC + 10 * ibranch;
+
+      int dtcselUpper = 0;
+      int dtcselLower = 0;
+      if (iRCU == 0) {
+        dtcselLower = (1 << ipos);
+      } else { // crate == 1
+        dtcselUpper = (1 << ipos);
+      }
+
+      for (ichip = 0; ichip < kNChip; ichip++) { // ALTRO 0,2,3,4
+        if (ichip != 1) {
+          for (ichan = 0; ichan < kNChan; ichan++) {
+            if (iFEC != 0 || (ichan < 8 || ichan > 11)) {
+              Ped = fMeanPed[iSM][iRCU][ibranch][iFEC][ichip][ichan];
+              int writeAddr = (ichip << 4) | ichan;
+              lineValue = FEClineCode | (writeAddr << 12) | Ped;
+              fout << lineValue << std::endl;
+            }
+          }
+        }
+      } // chip
+
+    } // iDTC
+  }   // iSM
+
+  //   std::vector<char> output;
+  std::string instructionString(fout.str());
+  //   std::fill(instructionString.begin(), instructionString.end(), std::back_inserter(output));
+  std::vector<char> output(instructionString.begin(), instructionString.end());
+  return output;
+}
+
+void EMCALPedestalHelper::dumpInstructions(const std::string_view filename, const std::vector<char> data, int mRun)
+{
+  std::ofstream fout(filename.data());
+  fout << mRun << "\n";
+  fout << data.data();
+  fout.close();
+}

--- a/Detectors/EMCAL/calibration/src/PedestalCalibDevice.cxx
+++ b/Detectors/EMCAL/calibration/src/PedestalCalibDevice.cxx
@@ -12,6 +12,7 @@
 #include "CommonUtils/VerbosityConfig.h"
 #include "DetectorsRaw/RDHUtils.h"
 #include "EMCALBase/Geometry.h"
+#include "Framework/TimingInfo.h"
 #include "EMCALCalibration/PedestalCalibDevice.h"
 #include "EMCALReconstruction/AltroDecoder.h"
 #include "EMCALReconstruction/RawReaderMemory.h"
@@ -39,6 +40,13 @@ void PedestalCalibDevice::init(o2::framework::InitContext& ctx)
 
 void PedestalCalibDevice::run(o2::framework::ProcessingContext& ctx)
 {
+  if (!mRun) {
+    const auto& tinfo = ctx.services().get<o2::framework::TimingInfo>();
+    if (tinfo.runNumber != 0) {
+      mRun = tinfo.runNumber;
+    }
+  }
+
   constexpr auto originEMC = o2::header::gDataOriginEMC;
   auto data = ctx.inputs().get<o2::emcal::PedestalProcessorData>(getPedDataBinding());
   LOG(debug) << "adding pedestal data";
@@ -60,6 +68,21 @@ void PedestalCalibDevice::sendData(o2::framework::EndOfStreamContext& ec, const 
 
   ec.outputs().snapshot(o2::framework::Output{o2::calibration::Utils::gDataOriginCDBPayload, "EMC_PEDCALIB", 0}, *image.get());
   ec.outputs().snapshot(o2::framework::Output{o2::calibration::Utils::gDataOriginCDBWrapper, "EMC_PEDCALIB", 0}, objInfo);
+
+  // the following goes to the DCS ccdb
+  EMCALPedestalHelper helper;
+  std::vector<char> vecPedData = helper.createPedestalInstruction(data);
+  if (mDumpToFile) {
+    helper.dumpInstructions("EMCAL-Pedestals.txt", vecPedData, mRun);
+  }
+
+  auto clNameDCS = o2::utils::MemFileHelper::getClassName(vecPedData);
+  auto flNameDCS = o2::ccdb::CcdbApi::generateFileName(clNameDCS);
+  o2::ccdb::CcdbObjectInfo objInfoDCS("EMC/Calib/PDData", clNameDCS, flNameDCS, md, mStartTS, o2::ccdb::CcdbObjectInfo::INFINITE_TIMESTAMP, true);
+  auto imageDCS = o2::ccdb::CcdbApi::createObjectImage(&vecPedData, &objInfoDCS);
+
+  ec.outputs().snapshot(o2::framework::Output{o2::calibration::Utils::gDataOriginCDBPayload, "EMC_PEDCALIBSTR", 1}, *imageDCS.get());
+  ec.outputs().snapshot(o2::framework::Output{o2::calibration::Utils::gDataOriginCDBWrapper, "EMC_PEDCALIBSTR", 1}, objInfoDCS);
 }
 
 //________________________________________________________________
@@ -72,7 +95,7 @@ void PedestalCalibDevice::endOfStream(o2::framework::EndOfStreamContext& ec)
   resetStartTS();
 }
 
-o2::framework::DataProcessorSpec o2::emcal::getPedestalCalibDevice()
+o2::framework::DataProcessorSpec o2::emcal::getPedestalCalibDevice(bool dumpToFile)
 {
 
   std::vector<o2::framework::InputSpec> inputs;
@@ -81,10 +104,13 @@ o2::framework::DataProcessorSpec o2::emcal::getPedestalCalibDevice()
   outputs.emplace_back(o2::framework::ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBPayload, "EMC_PEDCALIB"}, o2::framework::Lifetime::Sporadic);
   outputs.emplace_back(o2::framework::ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBWrapper, "EMC_PEDCALIB"}, o2::framework::Lifetime::Sporadic);
 
+  outputs.emplace_back(o2::framework::ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBPayload, "EMC_PEDCALIBSTR"}, o2::framework::Lifetime::Sporadic);
+  outputs.emplace_back(o2::framework::ConcreteDataTypeMatcher{o2::calibration::Utils::gDataOriginCDBWrapper, "EMC_PEDCALIBSTR"}, o2::framework::Lifetime::Sporadic);
+
   return o2::framework::DataProcessorSpec{
     "PedestalCalibrator",
     inputs,
     outputs,
-    o2::framework::AlgorithmSpec{o2::framework::adaptFromTask<o2::emcal::PedestalCalibDevice>()},
+    o2::framework::AlgorithmSpec{o2::framework::adaptFromTask<o2::emcal::PedestalCalibDevice>(dumpToFile)},
     o2::framework::Options{}};
 }

--- a/Detectors/EMCAL/calibration/testWorkflow/emc-pedestal-calib-workflow.cxx
+++ b/Detectors/EMCAL/calibration/testWorkflow/emc-pedestal-calib-workflow.cxx
@@ -31,7 +31,8 @@ using namespace o2::emcal;
 void customize(std::vector<ConfigParamSpec>& workflowOptions)
 {
   std::vector<ConfigParamSpec> options{
-    {"configKeyValues", VariantType::String, "", {"Semicolon separated key=value strings"}}};
+    {"configKeyValues", VariantType::String, "", {"Semicolon separated key=value strings"}},
+    {"dumpToFile", VariantType::Bool, false, {"if output (that goes to DCS ccdb) should be stored in txt file for local debugging"}}};
   std::swap(workflowOptions, options);
 }
 
@@ -41,9 +42,10 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
 {
 
   o2::conf::ConfigurableParam::updateFromString(cfgc.options().get<std::string>("configKeyValues"));
+  bool dumpToFile = cfgc.options().get<bool>("dumpToFile");
 
   WorkflowSpec specs;
-  specs.emplace_back(o2::emcal::getPedestalCalibDevice());
+  specs.emplace_back(o2::emcal::getPedestalCalibDevice(dumpToFile));
 
   return specs;
 }

